pkg/importer/dataStream.go

```diff
@@ -172,97 +181,151 @@ func CopyImage(dest, endpoint, accessKey, secKey string) error {
 	return ds.copy(dest)
 }
 
-// Parse the endpoint extensions and set the Reader slice in the receiver. The reader order starts with
-// the lowest level reader, eg. http, used to read file content. The last reader is a multi reader needed
-// to piece together a byte reader which is always used to read the magic number of the file. In between
-// are the decompression/archive readers, if any. Readers are closed in reverse (right-to-left) order,
+// Read the endpoint and determine the file composition (eg. .iso.tar.gz) based on the magic number in
+// each known file format header. Set the Reader slice in the receiver and set the Size field to each
+// reader's original size. Note: the reader processed last defines the final Size.
+// The reader order starts with the lowest level reader, eg. http, used to read file content. The next
+// readers are combinations of decompression/archive readers and bytes multi-readers. The multi-readers
+// are created so that header data (interpreted by the current reader) is present for the next reader.
+// Thus, the last reader in the reader stack is always a multi-reader. Readers are closed in reverse order,
 // see the Close method. If a format doesn't natively support Close() a no-op Closer is wrapped around
 // the native Reader so that all Readers can be consider ReadClosers.
 // Examples:
-//   Filename                         Readers
-//   --------                         -------
-//   "https:/somefile.tar.gz"         [http, gz, tar, multi-reader]
-//   "file:/somefile.tar.xz"          [file, xz, tar, multi-reader]
-//   "s3://some-iso"                  [s3, multi-reader]
-//   "https://somefile.qcow2"         [http, multi-reader]
-//   "https://somefile.qcow2.tar.gz"  [http, gz, tar, multi-reader]
-// Note: the .qcow2 ext is ignored; qemu files are detected based on their magic number.
+//   Filename                    Readers (mr == multi-reader)
+//   --------                    ----------------------------
+//   "https:/foo.tar.gz"         [http, mr, gz, mr, tar, mr]
+//   "file:/foo.tar.xz"          [file, mr, xz, mr, tar, mr]
+//   "s3://foo-iso"              [s3, mr]
+//   "https://foo.qcow2"         [http, mr]		     note: there is no qcow2 reader
+//   "https://foo.qcow2.tar.gz"  [http, mr, gz, mr, tar, mr] note: there is no qcow2 reader
+// Note: file extensions are ignored.
 // Note: readers are not closed here, see dataStream.Close().
+// Assumption: a particular header format only appears once in the data stream. Eg. foo.gz.gz is not supported.
 func (d *dataStream) constructReaders() error {
```

Comment:
Per bjeans convo - `constructReaders` over abstracts it's operation.  At a glance, it's not obvious that a reader slice is being build by `dataStreamSelector` and the format readers.  

The function should be easy understand without having to delve into it's subprocesses.   Specifically, `append()` should be called at this scope, not the scope of the subprocesses.

Reply:
@davidvossel 
> I see that each reader is inspecting the header, but it isn't obvious to me how the reader chain is processed during the copy phase. For the copy, I only see the top level reader (http/s3) being copied directly into a file, and then the qemu conversion happens after that. It isn't obvious to me were something like gzip, tar, etc... fits into the copy process.

The slice of readers is treated conceptually as a stack, with the client reader (http, s3) at the bottom and the final decoding reader at the top.  The `topReader()` method gets the top of the reader stack and passes it to the built in `copy()` method.  

When `copy()` calls `Read()` on the passed in reader, that call is propagated down the stack to the client reader.  The client reader begins streaming data, which is sent back up the stack and output by the top reader, then written to disk.

QCOW2 throws a wrench into this design.  The data cannot be operated on sequentially, so we cannot manipulate it in stream.  We're stuck having to write the entire file to disk, then use the qemu-img tool to convert the file to a raw format.  

One draw back to this a pretty clear: we're effectively, if temporarily, doubling (roughly) the storage footprint.  If we write into the container's root fs, there's a concern we'll run up against size limits enforced by the underlying node or the kubelet.  One thought has been to create a scratch PVC to write to, then deprovision it after.  Still not a great solution...